apps/emqx_ft/src/emqx_ft_storage_exporter_fs.erl

```diff
@@ -265,30 +293,35 @@ start_reader(Options, RelFilepath, CallerPid) ->
 
 %%
 
--spec list(options()) ->
+-spec list(options(), query()) ->
     {ok, [exportinfo(), ...]} | {error, [{node(), _Reason}]}.
-list(_Options) ->
+list(_Options, Query) ->
```

Comment:
Seems that much complexity comes from using the same function both for iteration and individual transfer fetching. Also, having N nodes, we seem to make (N-1)x extra RPC calls (if I understand correctly) to perform the listing.

But we are not forced to do this; here are some considerations:

* To get files for a transfer, we **must** make a multicall because files can be on any node. But in this case, we may suffice a simple glob lookup of a concrete folder (is this true?).
* Without having to support transfer lookup with the listing, we are not forced to do a multicall then. So we may treat a node as a part of the cursor (like a top level of a path) and iterate over each node sequentially. If I am not mistaken, this should reduce the number of RPC calls N times.

Reply:
Reimplemented with sequential iteration over sorted list of cluster nodes. While we no longer need mergesort-related stuff, I'd say that overall it didn't become much simpler than it was before (when comparing against naive and wasteful implementation though). Also it made cursor opaque and its structure became more complex, which wasn't really needed before since currently both local filesystem and S3 support transparent cursors (which are also a bit more useful to an API user).